state/indexer/sink/psql/psql_test.go

```diff
@@ -0,0 +1,365 @@
+		Distinct().From(TableResultTx).
+		InnerJoin(join).
+		Where(fmt.Sprintf("hash = $1 AND chain_id = '%s'", chainID), fmt.Sprintf("%X", hash))
+
+	rows, err := sqlStmt.RunWith(db).Query()
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	if rows.Next() {
+		var txResult []byte
+		var txResultID, txid int
+		var h, cid string
+		err = rows.Scan(&txResult, &txResultID, &txid, &h, &cid)
+		if err != nil {
+			return nil, nil
+		}
+
+		msg := new(abci.TxResult)
+		err = proto.Unmarshal(txResult, msg)
+		if err != nil {
+			return nil, err
+		}
+
+		return msg, err
+	}
+
+	// No result
+	return nil, nil
+}
+
+func verifyTimeStamp(tb string) error {
+
+	// We assume the tx indexing time would not exceed 2 second from now
+	sqlStmt := sq.
+		Select(fmt.Sprintf("%s.created_at", tb)).
+		Distinct().From(tb).
+		Where(fmt.Sprintf("%s.created_at >= $1", tb), time.Now().Add(-2*time.Second))
+
+	rows, err := sqlStmt.RunWith(db).Query()
+	if err != nil {
+		return err
+	}
+
+	defer rows.Close()
+
+	if rows.Next() {
+		var ts string
+		err = rows.Scan(&ts)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	return errors.New("no result")
+}
+
+func verifyBlock(h int64) (bool, error) {
+	sqlStmt := sq.
+		Select("height").
+		Distinct().
+		From(TableEventBlock).
+		Where(fmt.Sprintf("height = %d", h))
+	rows, err := sqlStmt.RunWith(db).Query()
+	if err != nil {
+		return false, err
+	}
+
+	defer rows.Close()
+
+	if !rows.Next() {
+		return false, nil
+	}
+
+	sqlStmt = sq.
+		Select("type, height", "chain_id").
+		Distinct().
+		From(TableEventBlock).
+		Where(fmt.Sprintf("height = %d AND type = '%s' AND chain_id = '%s'", h, types.EventTypeBeginBlock, chainID))
+
+	rows, err = sqlStmt.RunWith(db).Query()
+	if err != nil {
+		return false, err
+	}
+
+	if !rows.Next() {
+		return false, nil
+	}
+
+	sqlStmt = sq.
+		Select("type, height").
+		Distinct().
+		From(TableEventBlock).
+		Where(fmt.Sprintf("height = %d AND type = '%s'", h, types.EventTypeEndBlock))
+	rows, err = sqlStmt.RunWith(db).Query()
+
+	if err != nil {
+		return false, err
+	}
+
+	return rows.Next(), nil
+}
+
+func setupDB(t *testing.T) (*dockertest.Pool, error) {
+	t.Helper()
+	pool, err := dockertest.NewPool(os.Getenv("DOCKER_URL"))
+
+	assert.NoError(t, err)
+
+	resource, err = pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: DriverName,
+		Tag:        "13",
+		Env: []string{
+			"POSTGRES_USER=" + user,
+			"POSTGRES_PASSWORD=" + password,
+			"POSTGRES_DB=" + dbName,
+			"listen_addresses = '*'",
+		},
+		ExposedPorts: []string{port},
+	}, func(config *docker.HostConfig) {
+		// set AutoRemove to true so that stopped container goes away by itself
+		config.AutoRemove = true
+		config.RestartPolicy = docker.RestartPolicy{
+			Name: "no",
+		}
+	})
+
+	assert.NoError(t, err)
+
+	// Set the container to expire in a minute to avoid orphaned containers
+	// hanging around
```

Comment:
I'm a bit worried about this pattern. We should use cleanup/defers/etc. to avoid leaking containers (but even that's not particularly reliable). 

Is it easier to run (these tests) to just assume that a postgres instance is running localy and run the tests against that rather than orchestrating docker here. 

I'm sorry if this is part of the design and I missed it.

Reply:
@tychoish thanks for the comments, about this part, do you mean run these tests with a mock postgres instance instead of the test within the docker with a real postgres instance? If yes, we couldn't verify the real postgres behavior through mocking the postgres and I was worry about the test correctness.

For leaking containers concerns, the dockertest package has a flag to remove the container automatically. https://github.com/tendermint/tendermint/pull/6411/files#diff-db93f01c930f8f49a60f5d775716249cba20b1e60ebcf7241957a98bf377d234R324
And to set `resource.Expire(60)` was referring to this part https://github.com/ory/dockertest#removing-old-containers

Reply:
I'm more in favor of using a real postgress, but not strongly if we think dockertest is the most reliable mechanism. 

Expire seems like a good option, but [from the implementation](https://github.com/ory/dockertest/blob/99759c4e4881d747a77b4b907a7a3a1543aa346f/dockertest.go#L211) seems like it requires that the coordinating process have not panic'ed when the timer succeeds and also (of course) that the process in the container terminate successfully, neither of which I feel very confident in relying on regularly. It's probably fine, in the end.